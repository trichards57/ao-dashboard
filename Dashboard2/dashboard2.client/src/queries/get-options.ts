import { queryOptions } from "@tanstack/react-query";
import { notFound } from "@tanstack/react-router";

export default function getOptions<T>(uri: string, queryKey: string[]) {
  return queryOptions({
    queryKey,
    queryFn: async () => {
      const response = await fetch(uri, {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      });

      if (!response.ok) {
        if (response.status === 404) {
          throw notFound();
        }
        throw new Error("Network error");
      }

      return response.json() as Promise<T>;
    },
    staleTime: 10 * 60 * 1000,
    throwOnError: true,
  });
}
