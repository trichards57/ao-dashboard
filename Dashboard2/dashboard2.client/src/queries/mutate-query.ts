import { useMutation, useQueryClient } from "@tanstack/react-query";
import { notFound } from "@tanstack/react-router";

export default function useUpdate(uri: string, queryKey: string[]) {
  const queryClient = useQueryClient();

  return function useUpdateInner<TBody>(id: string) {
    return useMutation({
      mutationFn: async (body: TBody) => {
        const response = await fetch(`${uri}/${id}`, {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify(body),
        });

        if (!response.ok) {
          if (response.status === 404) {
            // eslint-disable-next-line @typescript-eslint/no-throw-literal
            throw notFound();
          }
          throw new Error("Failed to update.");
        }
      },
      onSuccess: () => {
        queryClient.invalidateQueries({ queryKey });
      },
    });
  };
}
