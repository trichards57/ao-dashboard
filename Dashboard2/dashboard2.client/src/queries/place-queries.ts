import {
  QueryClient,
  queryOptions,
  useSuspenseQuery,
} from "@tanstack/react-query";
import { notFound } from "@tanstack/react-router";

export type Region =
  | "All"
  | "EastOfEngland"
  | "EastMidlands"
  | "London"
  | "NorthEast"
  | "NorthWest"
  | "SouthEast"
  | "SouthWest"
  | "WestMidlands";

export function districtOptions(region: Region) {
  return queryOptions({
    queryKey: ["districts", region],
    queryFn: async () => {
      if (region == "All") {
        return [];
      }

      const response = await fetch(`/api/places/${region}`, {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      });

      if (!response.ok) {
        if (response.status === 404) {
          throw notFound();
        }
        throw new Error("Failed to fetch districts.");
      }

      return response.json() as Promise<string[]>;
    },
    staleTime: 10 * 60 * 1000,
  });
}

export function hubOptions(region: Region, district: string) {
  return queryOptions({
    queryKey: ["hubs", region, district],
    queryFn: async () => {
      if (region == "All" || district.toUpperCase() == "ALL") {
        return [];
      }

      const response = await fetch(`/api/places/${region}/${district}`, {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      });

      if (!response.ok) {
        if (response.status === 404) {
          throw notFound();
        }
        throw new Error("Failed to fetch hubs.");
      }

      return response.json() as Promise<string[]>;
    },
    staleTime: 10 * 60 * 1000,
  });
}

export function useDistricts(region: Region) {
  return useSuspenseQuery(districtOptions(region));
}

export function useHubs(region: Region, district: string) {
  return useSuspenseQuery(hubOptions(region, district));
}

export function preloadDistricts(queryClient: QueryClient, region: Region) {
  return queryClient.ensureQueryData(districtOptions(region));
}

export function preloadHubs(
  queryClient: QueryClient,
  region: Region,
  district: string,
) {
  return queryClient.ensureQueryData(hubOptions(region, district));
}
