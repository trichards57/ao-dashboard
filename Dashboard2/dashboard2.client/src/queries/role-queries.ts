import { QueryClient, useSuspenseQuery } from "@tanstack/react-query";

import getOptions from "./get-options";
import useUpdate from "./mutate-query";

export type Permission = "Deny" | "Read" | "Write";

export interface RolePermissions {
  id: string;
  name: string;
  permissions: Permission;
  vehicleConfiguration: Permission;
  vorData: Permission;
}

export interface RolePermissionsUpdate {
  permissions: Permission;
  vehicleConfiguration: Permission;
  vorData: Permission;
}

export function useUpdateRole(id: string) {
  return useUpdate("/api/roles", ["role"])<RolePermissionsUpdate>(id);
}

export function roleOptions(id: string) {
  return getOptions<RolePermissions>(`/api/roles/${id}`, ["role", id]);
}

export const allRoleOptions = getOptions<RolePermissions[]>("/api/roles", [
  "role",
]);

export function useRole(id: string) {
  return useSuspenseQuery(roleOptions(id));
}

export function useRoles() {
  return useSuspenseQuery(allRoleOptions);
}

export function preloadRole(queryClient: QueryClient, id: string) {
  return queryClient.ensureQueryData(roleOptions(id));
}

export function preloadRoles(queryClient: QueryClient) {
  return queryClient.ensureQueryData(allRoleOptions);
}
