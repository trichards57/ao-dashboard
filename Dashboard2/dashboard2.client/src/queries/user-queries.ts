import {
  useMutation,
  useQueryClient,
  UseSuspenseQueryOptions,
} from "@tanstack/react-query";
import { notFound } from "@tanstack/react-router";

export interface UserInfo {
  realName: string;
  userId: string;
  email: string;
  role: string;
  amrUsed: string;
  lastAuthenticated: string;
  otherClaims: Record<string, string>;
}

export interface UserWithRole {
  id: string;
  name: string;
  roleId: string;
  role: string;
}

export interface UserRoleUpdate {
  roleId: string;
}

export const useUpdateUser = (id: string) => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (user: UserRoleUpdate) => {
      const response = await fetch(`/api/users/${id}`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(user),
      });

      if (!response.ok) {
        if (response.status === 404) {
          throw notFound();
        }
        throw new Error("Failed to update user.");
      }
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["user"] });
    },
  });
};

export const userSettings: (
  id: string,
) => UseSuspenseQueryOptions<UserWithRole> = (id: string) => ({
  queryKey: ["user", id],
  queryFn: async () => {
    const response = await fetch(`/api/users/${id}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      if (response.status === 404) {
        throw notFound();
      }
      throw new Error("Failed to fetch user settings.");
    }

    return response.json() as Promise<UserWithRole>;
  },
  staleTime: 10 * 60 * 1000
});

export const allUserOptions: UseSuspenseQueryOptions<UserWithRole[]> = {
  queryKey: ["user"],
  queryFn: async () => {
    const response = await fetch("/api/users", {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      throw new Error("Failed to fetch users.");
    }

    return response.json() as Promise<UserWithRole[]>;
  },
  staleTime: 10 * 60 * 1000
};

export const userMeOptions: UseSuspenseQueryOptions<UserInfo | null> = {
  queryKey: ["me"],
  queryFn: async () => {
    const response = await fetch("/api/users/me", {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      return Promise.resolve(null);
    }

    return response.json() as Promise<UserInfo>;
  },
  staleTime: 1000 * 60 * 5,
};
