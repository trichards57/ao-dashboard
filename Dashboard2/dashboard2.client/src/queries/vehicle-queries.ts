import { useMutation, useQueryClient } from "@tanstack/react-query";

export interface VehicleSettings {
  id: string;
  registration: string;
  callSign: string;
  hub: string;
  district: string;
  region: string;
  vehicleType: string;
  forDisposal: boolean;
}

export interface UpdateVehicleSettings {
  registration: string;
  hub: string;
  callSign: string;
  district: string;
  region: string;
  vehicleType: string;
  forDisposal: boolean;
}

export const useUpdateVehicle = () => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (vehicle: UpdateVehicleSettings) => {
      const response = await fetch("/api/vehicles", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(vehicle),
      });

      if (!response.ok) {
        throw new Error("Failed to update vehicle.");
      }
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["vehicle"] });
    },
  });
};

export const vehicleSettings = (id: string) => ({
  queryKey: ["vehicle", id],
  queryFn: async () => {
    const response = await fetch(`/api/vehicles/${id}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      throw new Error("Failed to fetch vehicle settings.");
    }

    return response.json() as Promise<VehicleSettings>;
  },
});

export const settingsOptions = (
  region: string | undefined,
  district: string | undefined,
  hub: string | undefined,
) => ({
  queryKey: ["settings", region ?? "All", district ?? "All", hub ?? "All"],
  queryFn: async () => {
    const response = await fetch(
      `/api/vehicles?region=${region ?? "All"}&district=${district ?? "All"}&hub=${hub ?? "All"}`,
      {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      },
    );

    if (!response.ok) {
      throw new Error("Failed to fetch vehicle settings.");
    }

    return response.json() as Promise<VehicleSettings[]>;
  },
});
