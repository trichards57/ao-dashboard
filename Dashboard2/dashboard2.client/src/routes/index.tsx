import { createFileRoute } from "@tanstack/react-router";

import useTitle from "../components/useTitle";
import { redirectIfLoggedIn } from "../support/check-logged-in";

export function Index() {
  useTitle();

  return (
    <section className="hero">
      <div className="hero-body">
        <p className="title">Welcome</p>
        <p className="subtitle">
          Welcome to the AO Ambulance Dashboard. You will need to log in to use
          this application.
        </p>
      </div>
      <div className="hero-foot is-flex is-justify-content-center">
        <a className="button is-primary" href="/Identity/Account/Login">
          Log In
        </a>
      </div>
    </section>
  );
}

export const Route = createFileRoute("/")({
  beforeLoad: ({ context }) => redirectIfLoggedIn(context),
  component: Index,
});
