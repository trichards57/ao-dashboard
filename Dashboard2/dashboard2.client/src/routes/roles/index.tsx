import { createFileRoute, Link } from "@tanstack/react-router";
import { preloadRoles, useRoles } from "../../queries/role-queries";
import { useTitle } from "../../components/useTitle";
import {
  redirectIfLoggedOut,
  redirectIfNoPermission,
} from "../../support/check-logged-in";

export function Roles() {
  const { data } = useRoles();
  useTitle("User Roles");

  const roles = [...data].sort((a, b) => a.name.localeCompare(b.name));

  return (
    <>
      <h1 className="title">User Roles</h1>

      <table className="table is-striped is-fullwidth">
        <thead>
          <tr>
            <th>Role</th>
            <th>Vehicle Configuration</th>
            <th>VOR Data</th>
            <th>User Role</th>
            <th>Role Permissions</th>
            <th></th>
          </tr>
        </thead>
        <tbody>
          {roles.map((r) => (
            <tr key={r.id}>
              <td>{r.name}</td>
              <td>{r.vehicleConfiguration}</td>
              <td>{r.vorData}</td>
              <td>{r.permissions}</td>
              <td>{r.name == "Administrator" ? "Write" : "Deny"}</td>
              <td className="edit">
                {r.name != "Administrator" && (
                  <Link to="/roles/edit/$id" params={{ id: r.id }}>
                    Edit
                  </Link>
                )}
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </>
  );
}

export const Route = createFileRoute("/roles/")({
  component: Roles,
  loader: ({ context }) => preloadRoles(context.queryClient),
  beforeLoad: ({ context }) => {
    redirectIfLoggedOut(context);
    redirectIfNoPermission(context.canEditRoles);
  },
});
