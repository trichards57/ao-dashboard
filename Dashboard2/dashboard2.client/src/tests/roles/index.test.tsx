import { beforeEach, describe, expect, it, Mock, vi } from "vitest";
import ReactDOM from "react-dom/client";
import { act, render } from "@testing-library/react";
import { Roles } from "../../routes/roles";
import { RolePermissions, useRoles } from "../../queries/role-queries";

vi.mock("../../queries/role-queries");

const testRoles: RolePermissions[] = [
  {
    id: "a",
    name: "Administrator",
    permissions: "Write",
    vehicleConfiguration: "Deny",
    vorData: "Deny",
  },
  {
    id: "b",
    name: "Role 1",
    permissions: "Deny",
    vehicleConfiguration: "Read",
    vorData: "Read",
  },
  {
    id: "c",
    name: "Administrator",
    permissions: "Read",
    vehicleConfiguration: "Write",
    vorData: "Write",
  },
];

describe("Roles Component", () => {
  beforeEach(() => {
    (useRoles as Mock).mockReturnValue({
      data: testRoles,
    });
  });

  it("renders without error", () => {
    const root = ReactDOM.createRoot(document.createElement("div"));
    expect(() => {
      act(() => {
        root.render(<Roles />);
        root.unmount();
      });
    }).not.toThrow();
  });

  it("renders correctly", () => {
    const { asFragment } = render(<Roles />);
    expect(asFragment()).toMatchSnapshot();
  });
});
