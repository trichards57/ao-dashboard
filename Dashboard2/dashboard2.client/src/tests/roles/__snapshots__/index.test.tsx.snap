// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Roles Component > renders correctly 1`] = `
<DocumentFragment>
  <h1
    class="title"
  >
    User Roles
  </h1>
  <table
    class="table is-striped is-fullwidth"
  >
    <thead>
      <tr>
        <th>
          Role
        </th>
        <th>
          Vehicle Configuration
        </th>
        <th>
          VOR Data
        </th>
        <th>
          User Role
        </th>
        <th>
          Role Permissions
        </th>
        <th />
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Administrator
        </td>
        <td>
          Deny
        </td>
        <td>
          Deny
        </td>
        <td>
          Edit
        </td>
        <td>
          Write
        </td>
        <td
          class="edit"
        />
      </tr>
      <tr>
        <td>
          Administrator
        </td>
        <td>
          Edit
        </td>
        <td>
          Edit
        </td>
        <td>
          Read
        </td>
        <td>
          Write
        </td>
        <td
          class="edit"
        />
      </tr>
      <tr>
        <td>
          Role 1
        </td>
        <td>
          Read
        </td>
        <td>
          Read
        </td>
        <td>
          Deny
        </td>
        <td>
          Deny
        </td>
        <td
          class="edit"
        >
          <a
            to="/roles/edit/b"
          >
            Edit
          </a>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
